lib/marc_cleanup/record_level.rb

```diff
@@ -1091,91 +1091,211 @@ def sparse_record?(record)
+      field.tag
+    elsif field.subfields.select { |subfield| subfield.code == '6' }.size != 1
+      field.tag
+    elsif field['6'] !~ /^[0-9]{3}-[0-9]+/
+      field.tag
+    else
+      field['6'].gsub(/^([0-9]{3})-.*$/, '\1')
+    end
+  end
+
+  def invalid_subfield_codes(field_info:, tag:)
+    errors = []
+    field_info[:field].subfields.group_by(&:code).each_key do |code|
+      next unless field_info[:schema][tag]['subfields'][code].nil?
+
+      error = "Invalid subfield code #{code} in instance #{field_info[:field_num]}"
+      errors << error
+    end
+    errors
+  end
+
+  def multiple_subfield_instances(field_info:, tag:)
+    field_info[:field].subfields.group_by(&:code).select do |code, subfields|
+      subfields.size > 1 &&
+        field_info[:schema][tag]['subfields'][code] &&
+        field_info[:schema][tag]['subfields'][code]['repeat'] == false
+    end
+  end
+
+  def non_repeatable_subfield_codes(field_info:, tag:)
+    errors = []
+    multiple_subfield_instances(field_info: field_info, tag: tag).each_key do |code|
+      error = "Non-repeatable subfield code #{code} repeated in instance #{field_info[:field_num]}"
+      errors << error
+    end
+    errors
+  end
+
+  def validate_subfields(field_info:, tag:)
+    errors = invalid_subfield_codes(field_info: field_info, tag: tag)
+    errors + non_repeatable_subfield_codes(field_info: field_info, tag: tag)
+  end
+
+  def validate_indicators(field_info:, tag:)
+    errors = []
+    unless field_info[:schema][tag]['ind1'].include?(field_info[:field].indicator1)
+      errors << "Invalid indicator1 value #{field_info[:field].indicator1} in instance #{field_info[:field_num]}"
+    end
+    unless field_info[:schema][tag]['ind2'].include?(field_info[:field].indicator2)
+      errors << "Invalid indicator2 value #{field_info[:field].indicator2} in instance #{field_info[:field_num]}"
+    end
+    errors
+  end
+
+  def validate_field_length(field_info)
+    return [] unless invalid_field_length?(field_info[:field])
+
+    ["Field length above 9,999 bytes in instance #{field_info[:field_num]} of #{field_info[:field].tag}"]
+  end
+
+  def validate_data_field(field_info)
+    tag = field_tag(field_info[:field])
+    return [] unless field_info[:schema][tag]
+    return [] if tag == '880'
+    return [] if ('001'..'009').include?(tag)
+
+    errors = validate_field_length(field_info)
+    errors += validate_indicators(field_info: field_info, tag: tag)
+    errors += validate_subfields(field_info: field_info, tag: tag)
+    errors.map! { |text| text + " linked to field tag #{tag}" } if field_has_link?(field_info[:field])
+    errors
+  end
+
+  def validate_data_fields(record:, schema:)
+    hash = {}
+    record.fields('010'..'999').each do |field|
+      field_num = record.fields(field.tag).index(field) + 1
+      field_errors = validate_data_field({ field: field, field_num: field_num, schema: schema })
+      next unless field_errors.size.positive?
```

Comment:
Potential simplification:

```suggestion
      next unless field_errors.any?
```